sql/rpl_replica_commit_order_manager.cc

```diff
@@ -197,6 +205,19 @@ bool Commit_order_manager::wait(Slave_worker *worker) {
   return false;
 }
 
+#ifdef WITH_WSREP
+void Commit_order_manager::wait_for_galera_turn(THD* thd) {
+  Slave_worker *worker = dynamic_cast<Slave_worker *>(thd->rli_slave);
+  if(!worker) return;
+
+  mysql_mutex_lock(&m_galera_seqno_mtx);
+  while (m_workers.front() != worker->id) {
+    mysql_cond_wait(&m_galera_seqno_cond, &m_galera_seqno_mtx);
+  }
+  mysql_mutex_unlock(&m_galera_seqno_mtx);
```

Comment:
Consider a scenario like this

The source(master) server executes the below 2 queries in parallel and INSERT is binlogged before DELETE. On replica, they get assigned to w1 and w2 respectively to execute parallely


w1:    INSERT INTO t1 (11, NULL);
w2:    DELETE FROM t1 WHERE c2 <= 3;

    On replica server, they are applied parallel by w1 and w2. And the deadlock is:
      w1                                        w2
    -------------------------                 --------------------------
     Applying INSERT                          Applying DELETE
                                              Gets the row locks first.
     Waits for w2 to release
     the row lock.
                                              Waits for w1 to prepare in wait_for_galera_turn().

In such a case, w1 will wait for w2 to release row locks and w2 will wait for w1 to prepare and we will have deadlocks.

Reply:
I debugged the code further, it looks `Commit_order_manager::wait_on_graph()` uses MDL's `timed_wait()` to wait for its turn. This is done because it can provide protection against ACL and FTWRL queries. The `mysql_cond_wait()` here does not provide any protection against ACL and FTWRL and it may result in deadlock.

Can you please evaluate the effects of calling `Commit_order_manager::wait_on_graph()`  here as well (it will be called again in `ha_commit_low()`) ?

Reply:
Fixed.